core/src/repair_service.rs

```diff
@@ -502,27 +510,64 @@ impl RepairService {
         max_repairs: usize,
         slot: Slot,
         duplicate_slot_repair_statuses: &HashMap<Slot, DuplicateSlotRepairStatus>,
+        vote_tracker: &Arc<VoteTracker>,
     ) {
+        let mut seed = [0u8; 32];
+        thread_rng().fill(&mut seed);
+        let rng = &mut ChaChaRng::from_seed(seed);
         let mut pending_slots = vec![slot];
         while repairs.len() < max_repairs && !pending_slots.is_empty() {
-            let slot = pending_slots.pop().unwrap();
-            if duplicate_slot_repair_statuses.contains_key(&slot) {
-                // These are repaired through a different path
-                continue;
+            pending_slots.retain(|slot| !duplicate_slot_repair_statuses.contains_key(slot));
+            let mut next_pending_slots = vec![];
+            let mut level_repairs = HashMap::new();
+            for slot in &pending_slots {
+                if let Some(slot_meta) = blockstore.meta(*slot).unwrap() {
+                    let new_repairs = Self::generate_repairs_for_slot(
+                        blockstore,
+                        *slot,
+                        &slot_meta,
+                        std::usize::MAX,
+                    );
+                    if !new_repairs.is_empty() {
+                        level_repairs.insert(*slot, new_repairs);
+                    }
+                    next_pending_slots.extend(slot_meta.next_slots);
+                }
             }
-            if let Some(slot_meta) = blockstore.meta(slot).unwrap() {
-                let new_repairs = Self::generate_repairs_for_slot(
-                    blockstore,
-                    slot,
-                    &slot_meta,
-                    max_repairs - repairs.len(),
-                );
-                repairs.extend(new_repairs);
-                let next_slots = slot_meta.next_slots;
-                pending_slots.extend(next_slots);
-            } else {
-                break;
```

Comment:
We removed this `else` condition because it should be impossible to reach right?